### PR TITLE
Cap the maximum reward per ECDSA keep operator per interval

### PR DESCRIPTION
**Pull request by eth-r**
Refs #446

Set a limit on the amount of KEEP a single ECDSA keep operator can receive in a single reward interval.

Use the withdrawal pattern to track allocated and withdrawn rewards. Rewards exceeding the cap will be returned to the unallocated pool.

### CODE REVIEW

File: solidity/contracts/ECDSARewards.sol
Comment (pdyraga) [inline solidity/contracts/ECDSARewards.sol:207]: We may have a little problem with upgrades. When finalizing the upgrade, we are allocating all past intervals and then, tokens are moved to the new contract. Rewards for past intervals are distributed later and we may find ourselves in a situtation when there is a lot of unallocated rewards on the previous rewards contract and there is an incetive to stay there.

What do you think? One option could be to disable any allocations after finalizing the upgrade and having a way to send all unallocated rewards to the new contract.

Comment (eth-r) [reply solidity/contracts/ECDSARewards.sol]: Good catch, the unallocated rewards line should be using `sendToNewRewards` instead, which takes care of this correctly.

Comment (eth-r) [reply solidity/contracts/ECDSARewards.sol]: Derp, I was looking at the wrong version of the rewards upgrade. We should do this yes:

> One option could be to disable any allocations after finalizing the upgrade and having a way to send all unallocated rewards to the new contract.

File: solidity/contracts/ECDSARewards.sol
Comment (benlongstaff) [inline solidity/contracts/ECDSARewards.sol:94]: should the cap be for the beneficiary address rather than the operator address? 

0xABF7335F324873Ccb1960C9cDb7278E627dd39eF
received the highest rewards out of all the beneficiaries (i believe 882,885 KEEP) by running 4 nodes and only staking 416,668 KEEP across them by juicing large BTC lot sizes. They would be unaffected by the change

0xBF62A4cf0e78b8799dcA0B481E4669a760C0f99F
will receive 669,583 KEEP from running one node, staking 9,293,734 KEEP and bonding 3,750 ETH and not juicing the keeps. they have the highest number of keeps still open. They would be affected by the change

File: solidity/contracts/ECDSARewards.sol
Comment (pdyraga) [reply solidity/contracts/ECDSARewards.sol]: Addressed in https://github.com/keep-network/keep-core/pull/2136 and later in 3616caf58f750839b45bf9a268d12912ef9cba61.

File: solidity/contracts/ECDSARewards.sol
Comment (pdyraga) [reply solidity/contracts/ECDSARewards.sol]: Addressed in https://github.com/keep-network/keep-ecdsa/pull/584/commits/35c47b5dfd6fdcc0572050b117b554899f95de84.

File: solidity/contracts/ECDSARewards.sol
Comment (pdyraga) [inline solidity/contracts/ECDSARewards.sol:127]: If we do it this way, only operator can withdraw. I think we should rather allow to pass the operator/beneficiary as a parameter.

File: solidity/contracts/ECDSARewards.sol
Comment (pdyraga) [inline solidity/contracts/ECDSARewards.sol:102]: Should we add two public functions allowing to fetch this information?

**Comment by pdyraga** [comment]
I fixed a problem in tests in https://github.com/keep-network/keep-ecdsa/pull/584/commits/7dc61ccae4c373b5340dd448c1cea32cf8744f72, now they are failing as expected.

**Comment by pdyraga** [comment]
... And added calls to `withdrawRewards` where applicable so they are now passing as expected.

File: solidity/contracts/ECDSARewards.sol
Comment (eth-r) [reply solidity/contracts/ECDSARewards.sol:102]: Added, plus another one for fetching the withdrawable amount

File: solidity/contracts/ECDSARewards.sol
Comment (eth-r) [reply solidity/contracts/ECDSARewards.sol]: Operator added as parameter; we can also switch to beneficiary if we like that better.

**Comment by dimpar** [comment]
Code looks good to me, but tests will be added in a separate PR.

File: solidity/contracts/ECDSARewards.sol
Comment (pdyraga) [reply solidity/contracts/ECDSARewards.sol]: I think the operator makes more sense 👌

File: solidity/contracts/ECDSARewards.sol
Comment (pdyraga) [inline solidity/contracts/ECDSARewards.sol:164]: We may consider renaming this variable to just `allocated` so that the linter is not getting crazy about tabulating the line.

**Comment by pdyraga** [comment]
The last three comments and then I am ready to merge this one and https://github.com/keep-network/keep-ecdsa/pull/587 soon after.

File: solidity/contracts/ECDSARewards.sol
Comment (pdyraga) [reply solidity/contracts/ECDSARewards.sol:96]: Well... we can actually leave it as-is for consistency. We'll verify this contract anyway.

File: solidity/contracts/ECDSARewards.sol
Comment (pdyraga) [reply solidity/contracts/ECDSARewards.sol]: Done in https://github.com/keep-network/keep-ecdsa/pull/584/commits/873cb9a177301a3529a317d79127131c37a991de

File: solidity/contracts/ECDSARewards.sol
Comment (pdyraga) [reply solidity/contracts/ECDSARewards.sol]: Done in https://github.com/keep-network/keep-ecdsa/pull/584/commits/8da20610ac4291e61a20d53a98e4fcb13856eda7

**Review by pdyraga — APPROVED** [review]
Need approval from either @eth-r or @dimpar before the merge since I was also committing here.

**Review by dimpar — APPROVED** [review]
@pdyraga your last changes look good

**Pull request merged**
